### PR TITLE
Stats 12/pcollections

### PR DESCRIPTION
**Pull request by sopel39**
This improves performance of join enumeration by factor >2x

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/cost/EnsureStatsMatchOutput.java [hunk 1]
```diff
@@ -29,16 +30,16 @@
     @Override
     public PlanNodeStatsEstimate normalize(PlanNode node, PlanNodeStatsEstimate estimate, Map<Symbol, Type> types)
     {
-        Map<Symbol, SymbolStatsEstimate> symbolSymbolStats = new HashMap<>();
-        estimate.getSymbolsWithKnownStatistics().stream()
-                .filter(node.getOutputSymbols()::contains)
-                .forEach(symbol -> symbolSymbolStats.put(symbol, estimate.getSymbolStatistics(symbol)));
+        PlanNodeStatsEstimate.Builder builder = buildFrom(estimate);
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/cost/EnsureStatsMatchOutput.java:33]: This seems like unrelated change. Move to separate commit please.

File: presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java [hunk 2]
```diff
@@ -83,15 +82,7 @@ public PlanNodeStatsEstimate mapOutputRowCount(Function<Double, Double> mappingF
...
-                .setSymbolStatistics(symbolStatistics.entrySet().stream()
-                        .collect(Collectors.toMap(
-                                Map.Entry::getKey,
-                                e -> {
-                                    if (e.getKey().equals(symbol)) {
-                                        return mappingFunction.apply(e.getValue());
-                                    }
-                                    return e.getValue();
-                                })))
+                .addSymbolStatistics(symbol, mappingFunction.apply(symbolStatistics.get(symbol)))
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java:85]: Yet another unrelated change?

**Review by fiedukow — APPROVED** [review]

**Comment by fiedukow** [comment]
Please close if it's merged

File: presto-main/src/main/java/com/facebook/presto/cost/EnsureStatsMatchOutput.java (hunk 1 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/EnsureStatsMatchOutput.java:33]: It is required since we now remove extra stats instead of building output stats from ground up

File: presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java (hunk 2 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java:85]: That is the core of this improvement: original map is not copied. Instead only one column is added to existing, immutable map. Pcollections PMap guarantees that such operation is performant
